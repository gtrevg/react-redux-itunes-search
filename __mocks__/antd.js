import React, { Component } from 'react';
import { mockComponent } from '../src/utils/testUtils';

export class Table extends Component {
	render() {
		return mockComponent('Antd Table', this.props);
	}
}

class Search extends Component {
	render() {
		return mockComponent('Antd Search', this.props);
	}
}

export const Input = { Search };
