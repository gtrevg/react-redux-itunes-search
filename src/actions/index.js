import axios from 'axios';

export const SEARCH_FETCH_SEARCH = 'SEARCH@FETCH_SEARCH';
export const SEARCH_FETCH_SEARCH_REQUEST = 'SEARCH@FETCH_SEARCH/REQUEST';
export const SEARCH_FETCH_SEARCH_SUCCESS = 'SEARCH@FETCH_SEARCH/SUCCESS';
export const SEARCH_FETCH_SEARCH_FAILURE = 'SEARCH@FETCH_SEARCH/FAILURE';

const URL_API_ITUNES = 'https://itunes.apple.com/search'

export const fetchSearch = (term) => {
  const url = `${URL_API_ITUNES}?media=music&term=${term.replace(' ', '+')}`;
  const meta = { term };
  return doFetch(SEARCH_FETCH_SEARCH, url, meta);
}

/*
 * Extracted from: https://github.com/reactjs/redux/issues/1676
 ***/
export const doFetch = (type, url, meta = {}) => {

    // Redux Thunk will inject dispatch here:
    return dispatch => {

        // Reducers may handle this to set a flag like isFetching
        dispatch({
            type: type + '/REQUEST',
            payload: {
                meta
            }
        })

        //The response function
        const theResponse = (response) => {
            dispatch({
                type: type + '/SUCCESS',
                payload: {
                    response,
                    meta
                }
            });
        }

        //The error function
        const theError = (error) => {
            dispatch({
                type: type + '/FAILURE',
                payload: {
                    error,
                    meta
                }
            });
        }

        // Perform the actual API call
        return axios.get(url).then(theResponse, theError);
    }
}
