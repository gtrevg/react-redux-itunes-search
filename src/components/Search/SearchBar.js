import React from 'react';
import { Input } from 'antd';

import './SearchBar.css';

const Search = Input.Search;

const SearchBar = () => {
  return (
    <div className="SearchBar">
      <Search
          placeholder="Search for artists, songs, albums, genre,... Whatever!"
          enterButton="Search"
          size="large"
          onSearch={value => console.log(value)} />
        </div>
  );
};

export default SearchBar;
