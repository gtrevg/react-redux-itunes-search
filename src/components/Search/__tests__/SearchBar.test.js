import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { Input } from 'antd';
import { Provider } from 'react-redux';

import SearchBar from '../SearchBar';
import { fetchSearch } from '../../../actions';
import store from '../../../config/mockStore';
import { SEARCH_FETCH_SEARCH_REQUEST } from '../../../actions';

describe('SearchBar component', () => {
	const initialEntries = ['/'];
	const component = (
		<Provider store={store}>
			<SearchBar />
		</Provider>
	);

	it('Renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(component, div);
		ReactDOM.unmountComponentAtNode(div);
	});

	it('Snapshot matchs', () => {
		const tree = renderer.create(component).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('Check that the search button launchs an action', () => {
		const tree = TestUtils.renderIntoDocument(component);
		const search = TestUtils.findRenderedComponentWithType(tree, Input.Search);
		const { onSearch } = search.props;

		//Call the function
		const term = 'Michael Jackson';
		onSearch(term);

		const expectedActions = [
			{
				type: SEARCH_FETCH_SEARCH_REQUEST,
				payload: {
					meta: {
						term,
					},
				},
			},
		];
		expect(store.getActions()).toEqual(expectedActions);
	});
});
