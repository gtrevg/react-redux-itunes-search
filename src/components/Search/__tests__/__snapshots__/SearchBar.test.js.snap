// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component Snapshot matchs 1`] = `
<div
  className="SearchBar"
>
  <mocked
    name="Antd Search"
    props={
      Object {
        "enterButton": "Search",
        "onSearch": [Function],
        "placeholder": "Search for artists, songs, albums, genre...",
        "size": "large",
      }
    }
  />
</div>
`;
